skopt/optimizer/optimizer.py

```diff
@@ -0,0 +1,263 @@
+        Points at which objective has been evaluated.
+    * `yi` [scalar]:
+        Values of objective at corresponding points in `Xi`.
+    * `models` [list]:
+        Regression models used to fit observations and compute acquisition
+        function.
+    * `space`
+        An instance of `skopt.space.Space`. Stores parameter search space used
+        to sample points, bounds, and type of parameters.
+
+    """
+    def __init__(self, dimensions, base_estimator,
+                 n_random_starts=10, acq_func="EI", acq_optimizer="lbfgs",
+                 random_state=None, n_points=10000, n_restarts_optimizer=5,
+                 xi=0.01, kappa=1.96, n_jobs=1):
+        # Arguments that are just stored not checked
+        self.acq_func = acq_func
+        self.rng = check_random_state(random_state)
+        self.kappa = kappa
+        self.n_points = n_points
+        self.n_restarts_optimizer = n_restarts_optimizer
+        self.xi = xi
+        self.space = Space(dimensions)
+        self.models = []
+        self.Xi = []
+        self.yi = []
+
+        self._check_arguments(dimensions, base_estimator,
+                              n_random_starts,
+                              acq_func, acq_optimizer,
+                              n_jobs)
+
+        self.parallel = Parallel(n_jobs=n_jobs)
+
+    def _check_arguments(self, dimensions, base_estimator,
+                         n_random_starts,
+                         acq_func, acq_optimizer,
+                         n_jobs):
+        """Check arguments for sanity."""
+        if not is_regressor(base_estimator):
+            raise ValueError(
+                "%s has to be a regressor." % base_estimator)
+        self.base_estimator = base_estimator
+
+        if n_random_starts < 0:
+            raise ValueError(
+                "Expected `n_random_starts` >= 0, got %d" % n_random_starts)
+        self._n_random_starts = n_random_starts
+
+        if acq_optimizer == "auto":
+            warnings.warn("The 'auto' option for the acq_optimizer will be "
+                          "removed in 0.4.")
+            acq_optimizer = "lbfgs"
+        self.acq_optimizer = acq_optimizer
+        if self.acq_optimizer not in ["lbfgs", "sampling"]:
+            raise ValueError(
+                "Expected acq_optimizer to be 'lbfgs' or 'sampling', "
+                "got %s" % acq_optimizer)
+
+    def ask(self):
+        """Suggest next point at which to evaluate the objective."""
+        if self._n_random_starts > 0:
+            self._n_random_starts -= 1
+            return self.space.rvs(random_state=self.rng)[0]
+
+        else:
+            if not self.models:
+                raise ValueError("Random evaluations exhausted and no "
+                                 "model has been fit.")
+            # return point computed from last call to tell()
+            return self._next_x
+
+    def tell(self, x, y, fit=True):
+        """Record an observation of the objective function.
```

Comment:
I prefer this, where the fitting happens feels like an implementation detail :-/